osxkeychain.go

```diff
@@ -176,43 +188,61 @@ func AddInternetPassword(pass *InternetPassword) error {
 //
 // Returns an error if the lookup was unsuccessful.
 func FindInternetPassword(pass *InternetPassword) (*InternetPassword, error) {
-	resp := InternetPassword{}
-	protocol := protocolTypeToC(pass.Protocol)
-	authtype := C.SecAuthenticationType(C.kSecAuthenticationTypeHTTPBasic)
-	var cpassword unsafe.Pointer
-	var cpasslen C.UInt32
+	// TODO: Check fields for UTF-8 encoding and size fitting in
+	// 32 bits.
+
+	serverName := C.CString(pass.ServerName)
+	defer C.free(unsafe.Pointer(serverName))
+
+	// TODO: Make optional.
+	securityDomain := C.CString(pass.SecurityDomain)
+	defer C.free(unsafe.Pointer(securityDomain))
+
+	accountName := C.CString(pass.AccountName)
+	defer C.free(unsafe.Pointer(accountName))
+
+	path := C.CString(pass.Path)
+	defer C.free(unsafe.Pointer(path))
+
+	protocol := C.uint(protocolTypeToC(pass.Protocol))
+
+	authtype := C.uint(authenticationTypeToC(pass.AuthType))
+
+	var passwordLength C.UInt32
+
+	var password unsafe.Pointer
+
 	var itemRef C.SecKeychainItemRef
 
 	errCode := C.SecKeychainFindInternetPassword(
 		nil, // default keychain
 		C.UInt32(len(pass.ServerName)),
-		C.CString(pass.ServerName),
+		serverName,
 		C.UInt32(len(pass.SecurityDomain)),
-		C.CString(pass.SecurityDomain),
+		securityDomain,
 		C.UInt32(len(pass.AccountName)),
-		C.CString(pass.AccountName),
+		accountName,
 		C.UInt32(len(pass.Path)),
-		C.CString(pass.Path),
+		path,
 		C.UInt16(pass.Port),
-		protocol,
-		authtype,
-		&cpasslen,
-		&cpassword,
+		C.SecProtocolType(protocol),
+		C.SecAuthenticationType(authtype),
+		&passwordLength,
+		&password,
 		&itemRef,
 	)
 
-	if errCode != C.noErr {
-		if err, exists := resultCodes[int(errCode)]; exists {
-			return nil, err
-		}
-		return nil, fmt.Errorf("Unmapped result code: %d", errCode)
+	if err := newKeychainError(errCode); err != nil {
+		return nil, err
 	}
+
 	defer C.CFRelease(C.CFTypeRef(itemRef))
-	defer C.SecKeychainItemFreeContent(nil, cpassword)
+	defer C.SecKeychainItemFreeContent(nil, password)
+
+	resp := InternetPassword{}
+	resp.Password = C.GoStringN((*C.char)(password), C.int(passwordLength))
 
-	cp2 := (**C.char)(cpassword)
-	buf := C.GoStringN(*cp2, C.int(cpasslen))
-	resp.Password = string(buf)
+	// TODO: Audit the code below.
```

Comment:
Okay, I took a closer look at that code. I found another memory error -- you're not supposed to free pointers returned by CFStringGetCStringPtr -- but otherwise it looks reasonable. (Although the port problem is weird -- I had no luck, either.)

Removed TODO.